File: javascript/src/televentas/envio.js
import { hoyUTC } from "./fechasUtil.js";

export class Envio {
  /**
   * @param {number} idEnvio
   * @param {string} direccionDestino
   * @param {Date | null} fechaSalida
   * @param {string} estado
   */
  constructor(idEnvio, direccionDestino, fechaSalida, estado) {
    this.idEnvio = idEnvio;
    this.direccionDestino = direccionDestino;
    this.fechaSalida = fechaSalida;
    this.estado = estado;
  }

  getIdEnvio() {
    return this.idEnvio;
  }

  getDireccionDestino() {
    return this.direccionDestino;
  }

  /** @returns {Date | null} */
  getFechaSalida() {
    return this.fechaSalida;
  }

  getEstado() {
    return this.estado;
  }

  programar() {
    this.estado = "programado";
  }

  despachar() {
    this.estado = "despachado";
    this.fechaSalida = hoyUTC();
  }

  entregar() {
    this.estado = "entregado";
  }
}

File: java/src/televentas/Envio.java
package televentas;

import java.time.LocalDate;
import java.util.Optional;

public class Envio {

    private final int idEnvio;
    private final String direccionDestino;
    private LocalDate fechaSalida;
    private String estado;

    public Envio(int idEnvio, String direccionDestino, LocalDate fechaSalida, String estado) {
        this.idEnvio = idEnvio;
        this.direccionDestino = direccionDestino;
        this.fechaSalida = fechaSalida;
        this.estado = estado;
    }

    public int getIdEnvio() {
        return idEnvio;
    }

    public String getDireccionDestino() {
        return direccionDestino;
    }

    public Optional<LocalDate> getFechaSalida() {
        return Optional.ofNullable(fechaSalida);
    }

    public String getEstado() {
        return estado;
    }

    public void programar() {
        this.estado = "programado";
    }

    public void despachar() {
        this.estado = "despachado";
        this.fechaSalida = LocalDate.now();
    }

    public void entregar() {
        this.estado = "entregado";
    }
}
